app/src/main/java/com/github/sdp/tarjetakuna/location/Location.kt

```diff
@@ -0,0 +1,107 @@
+package com.github.sdp.tarjetakuna.location
+
+import android.Manifest
+import android.content.Context
+import android.content.pm.PackageManager
+import android.location.LocationListener
+import android.location.LocationManager
+import android.util.Log
+import androidx.appcompat.app.AppCompatActivity
+import androidx.core.app.ActivityCompat
+import com.github.sdp.tarjetakuna.model.Coordinates
+
+object Location {
+
+
+    private var currentLocation: Coordinates = Coordinates(0.0, 0.0)
+
+    private var locationManager: LocationManager? = null
+
+    /**
+     * Check if the location permission is granted, and if not, request it
+     */
+    fun askForLocationPermission(context: AppCompatActivity) {
+        if (ActivityCompat.checkSelfPermission(
+                context,
+                android.Manifest.permission.ACCESS_COARSE_LOCATION
+            ) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(
+                context,
+                android.Manifest.permission.ACCESS_FINE_LOCATION
+            ) != PackageManager.PERMISSION_GRANTED
+        ) {
+            requestPermission(context)
+        }
+    }
+
+    /**
+     * Get the current location of the user if the permission is granted
+     */
+    fun captureCurrentLocation(context: AppCompatActivity) {
+        // Check if the location permission is granted, and if not, request it
+        if (ActivityCompat.checkSelfPermission(

```

Comment:
You should create a function "checkPermissionLocation" or something similar to remove the duplicate code.

Reply:
Either way, you can just call the _askForLocationPermission_ function instead of copying this code.

Reply:
The problem is that it is required to do the permission check if we want to call the location manager and if I merge the two functions it will call twice the permission request if the user refuses

Reply:
adding a variable "hasAlreadyAsked" does not really work either because the location manager requires explicitly an if else that has the form of the already implemented function, (if I put for example if(conditionCheck(context)) {...}else{locationManager ...} it wants the correct form if (ActivityCompat.checkSelfPermission...) else {locationManager...}.
So it will still duplicate the code

Reply:
I don't think I have understood what you mean. You should create a new private function to delete duplicate code for this :

```
if (ActivityCompat.checkSelfPermission(context,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(
                context,
                Manifest.permission.ACCESS_COARSE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) 
```

Reply:
That's what I meant, if I create a private function as you said, it does not work because the location manager requires this code snippet to be in the same function as it is (it does not accept a function call to this code snippet)

Reply:
So something like that :

```
    /**
     * Check if the location permission is granted, and if not, request it
     */
    fun askForLocationPermission(context: AppCompatActivity) {
        if (checkPermission(context)) {
            requestPermission(context)
        }
    }

    /**
     * Get the current location of the user if the permission is granted
     */
    fun captureCurrentLocation(context: AppCompatActivity) {
        // Check if the location permission is granted, and if not, request it
        if (!checkPermission(context)) {
            Log.i("Location", "Permission not granted")
            // Do nothing since the user rejected the permission
        } else {
            if (locationManager == null) {
                locationManager =
                    context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
            }
            locationManager!!.requestLocationUpdates(
                LocationManager.GPS_PROVIDER,
                0,
                0f,
                locationListener
            )
        }

    }

    /**
     * Check if the location permission is granted
     */
    private fun checkPermission(context: AppCompatActivity): Boolean {
        return ActivityCompat.checkSelfPermission(
            context,
            Manifest.permission.ACCESS_FINE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(
            context,
            Manifest.permission.ACCESS_COARSE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED
    }
```

would not work?

Reply:
Nope you can try putting it on Kotlin it will tell you that the things in checkPermission should be in captureCurrentLocation, the askForPermission would work tho

Reply:
Apparently it has to have the exact name "checkpermission" to be able to put it in another function, so I refactored it thanks!